docs/presentations/DB-write-speed/db-write-speed.slide

```diff
@@ -0,0 +1,501 @@
+    - expected
+    - but is there any space for improvement?
+
+
+## Benchmark results
+
+- not the benchmark is run in steps
+    - 10,000 messages in each step
+    - without DB cleanup between steps
+    - so DB is (slowly) being filled-in
+
+.image benchmarks.png
+
+
+
+## Write duration for 110,000 new unique reports
+
+- in steps with 10,000 blocks
+
+.image write_duration.png
+
+
+
+## Write speed
+
+- measured for 100,000 new unique reports
+
+.image write_speed_1.png
+
+
+
+## Write speed
+
+- measured for 100,000 new unique reports
+
+- approximation added
+
+.image write_speed_2.png
+
+
+
+## DB operations during writing new reports
+
+```
+INSERT OR REPLACE INTO report(org_id, cluster, report, reported_at, last_checked_at, kafka_offset, gathered_at)
+VALUES ($1, $2, $3, $4, $5, $6, $7);
+```
+
+- performed once per message (report)
+
+```
+DELETE FROM recommendation WHERE cluster_id = $1;
+INSERT INTO recommendation (org_id, cluster_id, rule_fqdn, error_key, rule_id, created_at) VALUES ...
+```
+
+- `INSERT` performed several times, once per rule hit
+
+```
+DELETE FROM rule_hit WHERE org_id = $1 AND cluster_id = $2;
+INSERT OR REPLACE INTO rule_hit(org_id, cluster_id, rule_fqdn, error_key, template_data)
+                       VALUES ($1, $2, $3, $4, $5);
+```
+
+- `INSERT` performed several times, once per rule hit
+
+
+
+## Usage of EXPLAIN and EXPLAIN ANALYSE statements
+
+- (for semi-full database)
+
+
+
+## UPSERT into report table
+
+```
+EXPLAIN ANALYZE INSERT INTO report(org_id, cluster, report, reported_at, last_checked_at, kafka_offset, gathered_at)
+VALUES (1, 123456, '', '2020-01-01', '2020-01-01', 0, '2020-01-01');
+
+                                            QUERY PLAN                                            
+
+--------------------------------------------------------------------------------------------------
+ Insert on report  (cost=0.00..0.01 rows=1 width=100) (actual time=25.302..25.303 rows=0 loops=1)
+   ->  Result  (cost=0.00..0.01 rows=1 width=100) (actual time=0.003..0.004 rows=1 loops=1)
+ Planning Time: 0.105 ms
+ Execution Time: 35.888 ms
+(4 rows)
+```
+
+
+
+## UPSERT into recommendation table
+
+```
+EXPLAIN ANALYZE DELETE FROM recommendation WHERE cluster_id = '94288c61-b7db-4d23-9f81-ee88d376970d';
+
+                                                   QUERY PLAN                                                    
+-----------------------------------------------------------------------------------------------------------------
+ Delete on recommendation  (cost=0.00..2794.57 rows=9 width=6) (actual time=9.539..9.542 rows=0 loops=1)
+   ->  Seq Scan on recommendation  (cost=0.00..2794.57 rows=9 width=6) (actual time=0.015..9.401 rows=5 loops=1)
+         Filter: ((cluster_id)::text = '94288c61-b7db-4d23-9f81-ee88d376970d'::text)
+         Rows Removed by Filter: 69995
+ Planning Time: 0.143 ms
+ Execution Time: 9.603 ms
+(6 rows)
+```
+
+- seq scan?!?!?!
```

Comment:
does this mean that `cluster_id` is not an index in the `recommendation` table?

Reply:
exactly :/

Reply:
The index for that table is:
```
"recommendation_pk" PRIMARY KEY, btree (org_id, cluster_id, rule_fqdn, error_key)
```
This doesn't convert `cluster_id` into an index too?

Reply:
It creates just one "compound index" for the tuple `(org_id, cluster_id, rule_fqdn, error_key)`, but not just for `cluster_id` (at least AFAIK)

Reply:
at least this is easy to test :)